frontend/document/credocument.lua

```diff
@@ -283,10 +287,132 @@ function CreDocument:updateColorRendering()
     end
 end
 
+function CreDocument:setHideNonlinearFlows(hide_nonlinear_flows)
+    if hide_nonlinear_flows ~= self.hide_nonlinear_flows then
+        self.hide_nonlinear_flows = hide_nonlinear_flows
+        self._document:setIntProperty("crengine.doc.nonlinear.pagebreak.force", self.hide_nonlinear_flows and 1 or 0)
+        self:cacheFlows()
+    end
+end
+
 function CreDocument:getPageCount()
     return self._document:getPages()
 end
 
+-- Whether the document has any non-linear flow to care about
+function CreDocument:hasNonLinearFlows()
+    return self._document:hasNonLinearFlows()
+end
+
+-- Whether non-linear flows (if any) will be hidden
+function CreDocument:hasHiddenFlows()
+    return self.flows[1] ~= nil
+end
+
+-- Get the next/prev page number, skipping non-linear flows,
+-- i.e. the next/prev page that is either in the current
+-- flow or in the linear flow (flow 0)
+-- If "page" is 0, these give the initial and final linear pages
+function CreDocument:getNextPage(page)
+    if self:hasHiddenFlows() then
+        if page < 0 then return 0 end
+        local flow = page > 0 and self:getPageFlow(page) or 0
+        for test_page=page+1, self:getPageCount() do
+            local test_page_flow = self:getPageFlow(test_page)
+            if test_page_flow == flow or test_page_flow == 0 then
+                return test_page
+            end
+        end
+        return 0
```

Comment:
No case for breaking that `for` loop before reaching getPageCount() ? (I guess you can't for flow 0, but just asking).

Reply:
I don't think so. You have to keep testing until you find flow 0 or the end of the document. Unless you know in advance that a) you have reached the end of the current flow (this we can know as soon as a different flow is found), and b) there are no remaining pages in flow 0 (this is the tricky part).

Reply:
But nothing that another cache item cannot fix.